src/Gaffer/Spreadsheet.cpp

```diff
@@ -173,13 +179,206 @@ class RowsMapScope : boost::noncopyable, public Context::SubstitutionProvider
 
+// Manages a data structure mapping between row name
+// and row, so we can implement `RowsPlug::row()` without
+// linear search.
+class Spreadsheet::RowsPlug::RowNameMap
+{
+
+	public :
+
+		RowNameMap( Spreadsheet::RowsPlug *rowsPlug )
+			:	m_rowsPlug( rowsPlug )
+		{
+			rowsPlug->childAddedSignal().connect( boost::bind( &RowNameMap::childAdded, this, ::_2 ) );
+			rowsPlug->childRemovedSignal().connect( boost::bind( &RowNameMap::childRemoved, this, ::_2 ) );
+			rowsPlug->parentChangedSignal().connect( boost::bind( &RowNameMap::parentChanged, this ) );
+		}
+
+		RowPlug *row( const string &rowName )
+		{
+			if( !m_plugSetConnection.connected() )
+			{
+				// If we can't track name changes, we must do
+				// a brute force update. This should basically never
+				// happen, so we put no effort into making something more
+				// performant.
+				updateRowNames();
+			}
+
+			const auto &index = m_map.get<1>();
+			auto range = index.equal_range( rowName );
+			if( range.first == index.end() )
+			{
+				return nullptr;
+			}
+			else if( range.first == range.second )
+			{
+				// Single row with the requested name.
+				// This is the common case we are optimising
+				// for.
+				return range.first->plug;
+			}
+			else
+			{
+				// Multiple rows with the requested name.
+				// We must return the first one.
+				size_t minIndex = std::numeric_limits<size_t>::max();
+				RowPlug *result = nullptr;
+				for( auto it = range.first; it != range.second; ++it )
+				{
+					/// \todo We need GraphComponent to provide constant-time
+					/// access to a child's index.
+					const auto childIt = find( m_rowsPlug->children().begin(), m_rowsPlug->children().end(), it->plug );
+					const size_t index = childIt - m_rowsPlug->children().begin();
+					if( index < minIndex )
+					{
+						minIndex = index;
+						result = it->plug;
+					}
+				}
+				return result;
+			}
+		}
+
+	private :
+
+		string rowName( const RowPlug *row )
+		{
+			auto namePlug = row->namePlug()->source<StringPlug>();
+			if( namePlug->direction() == Plug::Out )
+			{
+				return "__rowNameMap::computedNameSentinel__";
+			}
+			else
+			{
+				return row->namePlug()->getValue();
+			}
+		}
+
+		void childAdded( GraphComponent *child )
+		{
+			auto row = static_cast<RowPlug *>( child ); // Guaranteed by `RowsPlug::acceptsChild()`.
+			if( row != m_rowsPlug->defaultRow() )
+			{
+				m_map.insert( { row, rowName( row ) } );
+			}
+		}
+
+		void childRemoved( GraphComponent *child )
+		{
+			auto row = static_cast<RowPlug *>( child ); // Guaranteed by `RowsPlug::acceptsChild()`.
+			m_map.erase( row );
+		}
+
+		void parentChanged()
+		{
+			if( auto node = m_rowsPlug->node() )
+			{
+				m_plugSetConnection = node->plugSetSignal().connect( boost::bind( &RowNameMap::plugSet, this, ::_1 ) );
+				m_plugInputChangedConnection = node->plugInputChangedSignal().connect( boost::bind( &RowNameMap::plugInputChanged, this, ::_1 ) );
```

Comment:
I was wondering about this at the time I wrote it. There is something I like about always connecting a `plugSetSignal` to a `plugSet` method rather than cutting out the middle man, but it's a bit fluffy. Let me see if I can make a half convincing case for it :

- It's a bit more explicit to have the control flow for `updateRowName()` written out in `plugSet()` rather than being slightly obfuscated by `bind()`. Bind still feels like magic sometimes, so I quite like to keep them simple.
- It's easier to extend when `plugSet()` needs to do something else as well.
- It results in a uniform `fooSignal/foo` pairing throughout the codebase.

I'm not sure these are ironclad reasons, probably more a preference, so if you have a strong case for doing the opposite I'd be interested in hearing it.

Reply:
That all makes sense, I guess I just liked the readability in `bind` for these simple cases as it shows the logic without having to go look elsewhere. If/When they need to be different I'd split them at that point.

Don't really feel strongly about it though probably just historical habits from past lives. Better to keep consistency in gaffer.